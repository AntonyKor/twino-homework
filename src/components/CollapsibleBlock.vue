<template>
  <div :class="{ collapsed: isCollapsed }" class="collapsible">
    <header>
      <h3>{{ title }}</h3>
      <button @click="isCollapsed = !isCollapsed">
        Show {{ isCollapsed ? "more" : "less" }}
      </button>
    </header>
    <div class="content"><slot></slot></div>
  </div>
</template>

<script>
export default {
  props: {
    title: {
      type: String,
      default: "",
    },
  },

  data() {
    return { isCollapsed: true };
  },
};
</script>

<style scoped lang="scss">
.collapsible {
  padding: 30px 0;
  border-bottom: 1px solid #9e9fa2;
}

header {
  display: flex;
  justify-content: space-between;
  align-items: center;

  button {
    margin-left: 25px;
    position: relative;
    background: none;
    border: none;
    outline: none;
    color: #2e50b6;
    font-weight: bold;
    cursor: pointer;

    &:before {
      position: absolute;
      content: "";
      display: block;
      width: 10px;
      height: 3px;
      background-color: #2e50b6;
      top: 50%;
      left: -15px;
      transform: translateY(-50%);
    }
  }
}

.content {
  margin-top: 30px;
}

.collapsed {
  .content {
    height: 0px;
    margin-top: 0;
    overflow: hidden;
  }

  header button:after {
    position: absolute;
    content: "";
    display: block;
    width: 10px;
    height: 3px;
    background-color: #2e50b6;
    top: 50%;
    left: -15px;
    transform: translateY(-50%) rotate(0.25turn);
  }
}
</style>
