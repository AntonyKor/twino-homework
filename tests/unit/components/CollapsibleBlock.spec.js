import { shallowMount } from "@vue/test-utils";
import CollapsibleBlock from "@/components/CollapsibleBlock.vue";

describe("CollapsibleBlock", () => {
  it("should match snapshot without optional props", () => {
    const wrapper = shallowMount(CollapsibleBlock);

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should match snapshot with optional props", () => {
    const wrapper = shallowMount(CollapsibleBlock, {
      propsData: { title: "Title" },
      slots: { default: "<p>slot</p>" },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render title in header", () => {
    const wrapper = shallowMount(CollapsibleBlock, {
      propsData: { title: "Title" },
      slots: { default: "<p>slot</p>" },
    });

    expect(wrapper.find("header h3").text()).toEqual("Title");
  });

  it("should render slot in div.content", () => {
    const wrapper = shallowMount(CollapsibleBlock, {
      propsData: { title: "Title" },
      slots: { default: "<p>slot</p>" },
    });

    expect(wrapper.find("div > p").text()).toEqual("slot");
  });

  it("should add collapsed class to main at start", () => {
    const wrapper = shallowMount(CollapsibleBlock, {
      propsData: { title: "Title" },
      slots: { default: "<p>slot</p>" },
    });

    expect(wrapper.find(".collapsible").classes("collapsed")).toBeTruthy();
  });

  it("should render 'Show more' text in button in header at start", () => {
    const wrapper = shallowMount(CollapsibleBlock, {
      propsData: { title: "Title" },
      slots: { default: "<p>slot</p>" },
    });

    expect(wrapper.find("header button").text()).toEqual("Show more");
  });

  it("should remove collapsed class to main after click on button", async () => {
    const wrapper = shallowMount(CollapsibleBlock, {
      propsData: { title: "Title" },
      slots: { default: "<p>slot</p>" },
    });

    await wrapper.find("header button").trigger("click");

    expect(wrapper.find(".collapsible").classes("collapsed")).toBeFalsy();
  });

  it("should render 'Show less' text in button in header after click on button", async () => {
    const wrapper = shallowMount(CollapsibleBlock, {
      propsData: { title: "Title" },
      slots: { default: "<p>slot</p>" },
    });

    await wrapper.find("header button").trigger("click");

    expect(wrapper.find("header button").text()).toEqual("Show less");
  });
});
