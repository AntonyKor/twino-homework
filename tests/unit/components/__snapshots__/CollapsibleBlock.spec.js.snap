// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleBlock should match snapshot with optional props 1`] = `
<div
  class="collapsible collapsed"
>
  <header>
    <h3>
      Title
    </h3>
     
    <button>
      
      Show more
    
    </button>
  </header>
   
  <div
    class="content"
  >
    <p>
      slot
    </p>
  </div>
</div>
`;

exports[`CollapsibleBlock should match snapshot without optional props 1`] = `
<div
  class="collapsible collapsed"
>
  <header>
    <h3>
      
    </h3>
     
    <button>
      
      Show more
    
    </button>
  </header>
   
  <div
    class="content"
  />
</div>
`;
